redcap_bridge/project_building.py

```diff
@@ -0,0 +1,182 @@
+import pathlib
+
+import numpy as np
+import pandas as pd
+
+import redcap_bridge
+from redcap_bridge.project_validation import \
+    validate_project_against_template_parts
+from redcap_bridge.utils import map_header_json_to_csv
+
+template_dir = pathlib.Path(redcap_bridge.__file__).parent / 'template_parts'
+
+
+def build_project(project_csv, output_file=None):
+    """
+    Build a complete RedCap Instrument CSV from a set of template_parts and a
+    project csv file.
+
+    Parameters
+    ----------
+    project_csv: str
+        Filepath of the project csv file
+    
+    output_file: str, None
+        Filepath of the resulting, complete project csv (with inserted
+        template_parts. If None, the content is only returned and not saved.
+        Default: None
+
+    Returns
+    -------
+        (list) list containing the lines of the complete project definition
+        including the template content
+    """
+    output = []
+
+    if isinstance(project_csv, str):
+        project_csv = pathlib.Path(project_csv)
+
+    with open(project_csv) as f:
+        for line in f.readlines():
+            # if line only contains reference then include reference here
+            if line[0] == '{' and line[-2:] == '}\n' and not (',' in line):
+                template_name = pathlib.Path(line[1:-2]).with_suffix('.csv')
+                include_file = (template_dir / template_name).resolve()
+
+                if not include_file.exists():
+                    raise ValueError(f'Can not build project {project_csv}. '
+                                     f'Included template {include_file} does '
+                                     f'not exist')
+
+                with open(include_file, 'r') as f_include:
+                    header = f_include.readline()
+                    if header != output[0]:
+                        raise ValueError('Project csv and template do not share'
+                                         'the same header. Compare '
+                                         f'{project_csv} and {include_file}')
+
+                    output.extend(f_include.readlines())
+            else:
+                output.append(line)
+
+    if output_file:
+        with open(output_file, 'w') as f:
+            f.writelines(output)
+
+    return output
+
+
+def customize_project(project_built_csv, customization_csv, output_file=None):
+    """
+    Fill in a built project csv with project specific customizations.
+
+    This can be used to e.g. change the default values of fields or customize
+    the list of experimenters to be selected
+
```

Comment:
I think there is some kind of wrong indentation no  ?

Reply:
Yes I think I fixed it now.